### PR TITLE
UI/Card: Refactor Card component for improved accessibility

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
This PR refactors several aspects of the card component to better accommodate accessibility.
Main changes:
- The `heading` and `description` props have been factored out as card sub-components (`Card.Heading`, `Card.Description`), for better consistency with how other aspects of the card are defined (Meta, Tags, etc), and allowing for more flexibility.
- `className` props have been added to sub-components.
- Card component is now fully keyboard-navigable
- Sub-component order is respected. i.e. The order that sub-components are defined is reflected in the DOM. Note that this doesn't have any visual effect by default, but that can be changed using `grid-template-areas`.

### CODE REVIEW

File: packages/grafana-e2e-selectors/src/selectors/pages.ts [hunk 1]
```diff
@@ -29,7 +29,9 @@ export const Pages = {
   },
   AddDataSource: {
     url: '/datasources/new',
+    /** @deprecated */
     dataSourcePlugins: (pluginName: string) => `Data source plugin item ${pluginName}`,
+    dataSourcePlugins2: (pluginName: string) => `Add data source ${pluginName}`,
```
Comment (axelavargas) [inline packages/grafana-e2e-selectors/src/selectors/pages.ts:34]: I think we are using this other conversion based on this [PR](https://github.com/grafana/grafana/pull/41787):

```suggestion
    dataSourcePluginsV2: (pluginName: string) => `Add data source ${pluginName}`,
```

File: packages/grafana-ui/src/components/Badge/Badge.tsx [hunk 2]
```diff
@@ -25,7 +25,7 @@ export const Badge = React.memo<BadgeProps>(({ icon, color, text, tooltip, class
     <div className={cx(styles.wrapper, className)} {...otherProps}>
       <HorizontalGroup align="center" spacing="xs">
         {icon && <Icon name={icon} size="sm" />}
-        <span>{text}</span>
+        <span aria-label={tooltip}>{text}</span>
```
Comment (axelavargas) [inline packages/grafana-ui/src/components/Badge/Badge.tsx:28]: 🤔 Do you think here we should use `aria-labelledby` considering we have a visible text?

File: packages/grafana-ui/src/components/List/AbstractList.tsx [hunk 3]
```diff
@@ -31,11 +32,11 @@ export class AbstractList<T> extends React.PureComponent<AbstractListProps<T>> {
   }
 
   render() {
-    const { items, renderItem, getItemKey, className, inline } = this.props;
+    const { items, renderItem, getItemKey, className, inline, 'aria-labelledby': ariaLabelledBy } = this.props;
```
Comment (axelavargas) [inline packages/grafana-ui/src/components/List/AbstractList.tsx:35]: 🤔 I feel these changes don't belong in this PR, is it maybe a good idea to move it to another one, and keep the refactor scoped?

File: packages/grafana-ui/src/components/Tags/Tag.tsx [hunk 4]
```diff
@@ -17,26 +17,28 @@ export interface Props extends Omit<HTMLAttributes<HTMLElement>, 'onClick'> {
...
+    event.preventDefault();
+    event.stopPropagation();
+
+    onClick?.(name, event);
   };
 
+  const Wrapper = onClick ? 'button' : 'span';
   return (
-    <span
+    <Wrapper
```
Comment (axelavargas) [inline packages/grafana-ui/src/components/Tags/Tag.tsx:33]: Here the same comment that I did above applies :)

**Review by axelavargas — COMMENTED** [review]
Nice job in this refactor @kaydelaney 👏🏾 , I am wondering how do we handle breaking changes? I imagine someone using this component will need to refactor their code right?

Before

`<Card heading="Filter by name" description="Filter data by query." /> `

Now:

```
<Card>
  <Card.Heading>Filter by name</Card.Heading>
  <Card.Description>Filter data by query.</Card.Description>
</Card>
```

File: packages/grafana-e2e-selectors/src/selectors/pages.ts (hunk 1 above)
Comment (kaydelaney) [reply packages/grafana-e2e-selectors/src/selectors/pages.ts:34]: Ah, nice catch. 👍

File: packages/grafana-ui/src/components/List/AbstractList.tsx (hunk 3 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/List/AbstractList.tsx:35]: Haha, yeah that's fair, got a bit carried away.

**Comment by kaydelaney** [comment]
> Nice job in this refactor @kaydelaney 👏🏾 , I am wondering how do we handle breaking changes? I imagine someone using this component will need to refactor their code right?
> 
> Before
> 
> `<Card heading="Filter by name" description="Filter data by query." /> `
> 
> Now:
> 
> ```
> <Card>
>   <Card.Heading>Filter by name</Card.Heading>
>   <Card.Description>Filter data by query.</Card.Description>
> </Card>
> ```

Yeah, you're absolutely right. I asked what the deal with breaking changes and grafana-ui was in #frontend, but I think maybe my question slipped through the cracks. 😅 Any idea @jackw?

**Comment by jackw** [comment]
> Yeah, you're absolutely right. I asked what the deal with breaking changes and grafana-ui was in #frontend, but I think maybe my question slipped through the cracks. 😅 Any idea @jackw?

Sorry @kaydelaney for not replying sooner. Nice work on this! I'm not sure how wise it would be right now to put this out in a minor release. There's quite a few changes in here but the breaking changes that stood out to me relate to the replacement of `heading` and `description` props with components? Is there anything we can do to keep those props, mark them as deprecated, and introduce the additions you've made in the PR? The alternative might be to introduce a new Card component or  "list item" and deprecate this one? 🤔

File: packages/grafana-ui/src/components/Card/Card.tsx [hunk 5]
```diff
@@ -21,83 +18,181 @@ export interface Props extends Omit<CardContainerProps, 'disableEvents' | 'disab
...
+            return React.cloneElement(c, { disabled, styles, href, onClick: onCardClick, ...c.props });
+          } else {
+            return React.cloneElement(c, { disabled, styles, ...c.props });
+          }
+        }
+
+        return c;
+      }),
+    [children, disabled, href, styles, onCardClick]
+  );
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Card/Card.tsx:58]: If we're taking the time to refactor this component, I would like to take the chance to maybe take a refactor our approach here (which was present in the previous code).

This style of fiddling with child props, especially based on looking at the displayName seems off too me. It's not composable and not intuitive as to what'll happen I think.

For onClick and Heading, it looks like this is a change in behaviour? Previously `<Card onClick={} />` would be attached the the root, but now its put on just the title? Not knowing the motivations behind this, I think i would suggest to preserve the previous behaviour, and consumers can explicitly attach onClick to `<Card.Heading />` if they want that?

I think then each of the child components could then just call the styles function themselves to get them, and maybe `disabled` could  be handled with a css attribute selector?

For sharing values with children like this, where absolutely necessary I think it's better to use Context to send the values down rather than `.cloneElement`

File: packages/grafana-ui/src/components/Card/Card.tsx [hunk 6]
```diff
@@ -21,83 +18,181 @@ export interface Props extends Omit<CardContainerProps, 'disableEvents' | 'disab
...
+          {children}
+        </a>
+      ) : onClick ? (
+        <button className={styles?.linkHack} onClick={onClick} aria-label={ariaLabel}>
+          {children}
+        </button>
+      ) : (
+        <>{children}</>
+      )}
+    </h2>
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Card/Card.tsx:105]: I'm just not quite sure if this is the right way to produce simple and composable components. 

Ideally all `<Heading />` does is style it's children. I think then if they should be links or buttons, the consumer should pass an appropriate button or link in. I think this makes the components a bit more manageable? 

Thoughts?

File: packages/grafana-ui/src/components/Card/Card.tsx [hunk 7]
```diff
@@ -21,83 +18,181 @@ export interface Props extends Omit<CardContainerProps, 'disableEvents' | 'disab
...
     }
-  };
+    meta = filtered.reduce((prev, curr, i) => [
+      prev,
+      <span key={`separator_${i}`} className={styles?.separator}>
+        {separator}
+      </span>,
+      curr,
+    ]);
+  }
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Card/Card.tsx:153]: Imagine if we had a component that was like 

```
<Interpose node={separator}> 
  <A />
  <B />
</Interpose>
```

that could do this for us 😍 

(not a suggestion for now, just a mild thought for later 😄  )

File: packages/grafana-ui/src/components/Card/Card.tsx [hunk 8]
```diff
@@ -137,40 +228,58 @@ export const getCardStyles = stylesFactory((theme: GrafanaTheme2) => {
...
+      all: unset;
+      &::after {
+        position: absolute;
+        content: '';
+        top: 0;
+        bottom: 0;
+        left: 0;
+        right: 0;
+        border-radius: ${theme.shape.borderRadius(1)};
+      }
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Card/Card.tsx:256]: hmmmmmmm I see....

File: packages/grafana-ui/src/components/Tags/TagList.tsx [hunk 9]
```diff
@@ -7,33 +7,45 @@ export interface Props {
   onClick?: OnTagClick;
   /** Custom styles for the wrapper component */
   className?: string;
+  'aria-label'?: string | ((name: string, i: number) => string);
 }
 
-export const TagList: FC<Props> = memo(({ tags, onClick, className }) => {
+export const TagList: FC<Props> = memo(({ tags, onClick, className, 'aria-label': ariaLabel }) => {
```
Comment (joshhunt) [inline packages/grafana-ui/src/components/Tags/TagList.tsx:13]: When you move this to another PR 😉 , perhaps this prop should be renamed to like `getAriaLabel` to reduce confusion around "when can i pass a function or not?"?

**Review by joshhunt — COMMENTED** [review]
I love all the work and consideration that's went into this.

If we're going to do a significant refactor like this, I think we should take the approach to simplify (or remove 😅 ) how the `<Card />` component plays with the props of its children. 

Also, i think it's possible to make this change backwards compatible. Deprecate the `heading` and `description` props, and if they're present you can just insert the Card.Heading and Card.Description elements yourself?

I'm also not quite sure I understand the motivation for the  change to the link/onClick. I can see how they've been moved to the Heading and then a pseudo-element makes them cover the whole card (?). I guess this is so the link can be described by the Heading? Would it be better to use aria-labelledby on the card?

File: packages/grafana-ui/src/components/Card/Card.tsx (hunk 6 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Card/Card.tsx:105]: I agree with you. I guess I was trying to limit how much existing `Card` usage would need to be fixed up, but I suppose we might as well since it's a breaking change either way.

Comment (joshhunt) [reply packages/grafana-ui/src/components/Card/Card.tsx:105]: I think you could do both.

Preferred, non-deprecated way is:

```
<Card>
  <Card.Heading>
    <Link>...</Link>
  <Card.Heading/>
</Card>
```

And then for backwards compat you can also do:

```
<Card title="..." href="..." />
```

Where the implementation is something like:

```js
const Card = ({ title, href, children }) => {
    const headingCompat = href ? <a href={href}>{title}</a> : title

    return <div>
        {title &&
            <Card.Heading>
                {headingCompat}
            </Card.Heading>
        }
        {children}
    </div>
}
```

with title and href props marked as deprecated?

File: packages/grafana-ui/src/components/Card/Card.tsx (hunk 5 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Card/Card.tsx:58]: Ah, yeah, the reason for putting onClick/href on the heading vs the card container itself is pretty unintuitive - should've made a comment. The problem is that, if `href` or `onClick` is on the wrapper component itself, certain screen readers won't read out interactive components nested within that wrapper, which is of course the case a lot of the time with `Card` components. You could just put the `href` or `onClick` on the heading, but then you lose the whole card itself being clickable, which is a shame. The solution which I shamelessly stole from [here](https://inclusive-components.design/cards/) is to use pseudo-elements and some CSS trickery.

I don't mind the compound component pattern so much, but I do agree the implementation here could be better. Good point about using context

File: packages/grafana-ui/src/components/Card/Card.tsx (hunk 6 above)
Comment (kaydelaney) [reply packages/grafana-ui/src/components/Card/Card.tsx:105]: The only issue I have with this is that part of the reasoning of having `href` on the `Card` rather than the `Heading` was so the implementation detail of the `href` being on the `Heading` was elided, kind of. We want the card itself to be a link, but can't just slap the href on it because accessibility, but it can at least seem like it is by putting it on the heading together with some css tricks to cover the card. When it's something that the user passes to the heading directly, it might seem strange to then have the entire card be clickable, rather than _just_ the heading.

Comment (joshhunt) [reply packages/grafana-ui/src/components/Card/Card.tsx:105]: Hmmm yup okay I understand...

File: public/app/features/datasources/NewDataSourcePage.tsx [hunk 10]
```diff
@@ -131,37 +139,90 @@ const DataSourceTypeCard: FC<DataSourceTypeCardProps> = (props) => {
...
+      width: 'inherit',
+      marginRight: '0px',
+    }),
+    meta: css({
+      marginTop: '6px',
+      position: 'relative',
+    }),
+    description: css({
+      margin: '0px',
+      fontSize: '12px',
```
Comment (ashharrison90) [inline public/app/features/datasources/NewDataSourcePage.tsx:196]: can we get this from the theme?

File: public/app/features/datasources/NewDataSourcePage.tsx [hunk 11]
```diff
@@ -131,37 +139,90 @@ const DataSourceTypeCard: FC<DataSourceTypeCardProps> = (props) => {
...
+    heading: css({
+      fontSize: theme.v1.typography.heading.h5,
+      fontWeight: 'inherit',
+    }),
+    figure: css({
+      width: 'inherit',
+      marginRight: '0px',
+    }),
+    meta: css({
+      marginTop: '6px',
```
Comment (ashharrison90) [inline public/app/features/datasources/NewDataSourcePage.tsx:191]: can we get this from the theme? or does this specifically have to be 6px exactly? 🤔

File: public/app/features/datasources/NewDataSourcePage.tsx [hunk 12]
```diff
@@ -131,37 +139,90 @@ const DataSourceTypeCard: FC<DataSourceTypeCardProps> = (props) => {
...
+      "Figure    Meta     Actions"
+      "Figure     -       Actions"`,
+      backgroundColor: theme.colors.background.canvas,
+      border: '1px solid transparent',
+    }),
+    logo: css({
+      marginRight: theme.v1.spacing.lg,
+      marginLeft: theme.v1.spacing.sm,
+      width: '55px',
+      maxHeight: '55px',
```
Comment (ashharrison90) [inline public/app/features/datasources/NewDataSourcePage.tsx:219]: can we get/add this to the theme?

**Comment by torkelo** [comment]
Pushed a small update to remove the custom card background on the NewDataSourcePage , it's time has passed :) 
Screenshot from 2021-11-30 15-32-24

**Review by ashharrison90 — APPROVED** [review]
some very small nits. the behaviour seems really nice, and love that you've maintained backwards compatibility 🙌

**Review by Clarity-89 — COMMENTED** [review]
Great work with simplifying this component, especially using the context to handle the children components' props! 

One issue is that this breaks Enterprise as we use `CardContainer` component in the report list page. The idea of the `CardContainer` component is that it enables using our card styles for the cases where the Card component doesn't fit, mostly when the layout inside the card is too custom. Removing `href` prop from there makes it impossible to use is a wrapper for clickable list items, which are links. Maybe a workaround would be to wrap elements inside `CardContainer` into `a` tags directly 🤔

File: packages/grafana-ui/src/components/Card/Card.tsx [hunk 13]
```diff
@@ -1,225 +1,211 @@
...
-      margin: 0 ${theme.spacing(1)};
-    `,
-    tagList: css`
-      max-width: 50%;
-    `,
-  };
-});
+  return (
+    <h2 className={cx(styles.heading, className)}>
+      <Wrapper href={href} onClick={onClick} className={styles.linkHack} aria-label={ariaLabel}>
```
Comment (Clarity-89) [inline packages/grafana-ui/src/components/Card/Card.tsx:96]: I think this might need some conditional rendering as currently there's an error in console ```Warning: Invalid prop `onClick` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.``` I think even if href and/or onClick are undefined, they still trigger this error when passed to Fragment.

**Review by torkelo — COMMENTED** [review]
There might be some kinks to iron out and some implications for the enterprise reporting list, but overall these are really great changes, so hope we can resolve any issues! awesome work :rocket:

**Review by axelavargas — APPROVED** [review]
@kaydelaney Fantastic job in this refactor, it's really cool you were able to make it backward compatible :).

**Comment by kaydelaney** [comment]
So, bit of a decision to make here: While the interface for the card component (props, etc.) can be made backwards compatible, I suspect there's no easy way to make the actual styling backwards compatible since changing the implementation means css selectors previously used to change certain elements of the component won't work as expected anymore.
The question then is, even if the interface isn't broken, should it still be considered a breaking change since it can literally break how old components that use Card look? If so, should we just keep both the old Card, and the new one, and remove the old one for 9.0?

**Comment by axelavargas** [comment]
> So, bit of a decision to make here: While the interface for the card component (props, etc.) can be made backwards compatible, I suspect there's no easy way to make the actual styling backwards compatible since changing the implementation means css selectors previously used to change certain elements of the component won't work as expected anymore. The question then is, even if the interface isn't broken, should it still be considered a breaking change since it can literally break how old components that use Card look? If so, should we just keep both the old Card, and the new one, and remove the old one for 9.0?

Hello Kay, 🤔 my 2 cents here will be to keep the old `Card` and the new one, as you said, it might be the case that those CSS selectors are being used to override styles.

**Comment by torkelo** [comment]
I do not think we need to be backwards compatible with users override internal styling

**Comment by mckn** [comment]
> I do not think we need to be backwards compatible with users override internal styling

@torkelo can you elaborate a bit more on your thoughts here. I have probably missed something.

**Comment by torkelo** [comment]
@mckn I do not think overriding the Card styles is common, and if plugins do I I would not expect backwards compatibility. I mean if I use a component framework, and start overring component styles and internal div styles I would not expect those overrides to continue working in new versions

**Comment by torkelo** [comment]
But let's say we add NewCard and in v9 we make Card = NewCard, then if that is an API breaking change all plugins that used old Card (without overring styles) would break. But I think that is an alternative to the compatible API that could work as long as we remove the old component in next major

**Comment by mckn** [comment]
> @mckn I do not think overriding the Card styles is common, and if plugins do I I would not expect backwards compatibility. I mean if I use a component framework, and start overring component styles and internal div styles I would not expect those overrides to continue working in new versions

Yes, totally agree! I would consider the styling of the card to be an internal API of the component. If you override the styles I would say, sorry - but you are on your own. 

When it comes to the external API of the component, to me, it makes sense to keep backward compatibility and if needed remove the "old" component in next major version. What would make sense in this PR is to deprecate the "old" component and provide migration steps in the plugins migration guide so it is available to developers prior to v9 is released.

**Comment by torkelo** [comment]
If I understood it correctly the api could be made backward compatible except when users  override styles, which is why I expressed the thought that then we  might not need introduce a whole new component. But if you feel a whole new component is warranted and we can remove it in v9 then that is also an option

**Comment by mckn** [comment]
> If I understood it correctly the api could be made backward compatible except when users override styles, which is why I expressed the thought that then we might not need introduce a whole new component. But if you feel a whole new component is warranted and we can remove it in v9 then that is also an option

Aha, Sorry! So that was what I did not understand hence my initial question. Thanks for making this clear and as I said, I totally agree!

**Comment by kaydelaney** [comment]
Fairly confident this should be okay to merge now. Should be backwards compatible and not break anything for enterprise.
Any objections to merging? @torkelo @mckn

**Review by torkelo — APPROVED** [review]
